Feature/request order

alertjjm opened the pull request:
Feat: Redis 설정
- gradle에서 spring data redis 설정
- RedisConfig 설정
- application.properties에 레디스 정보 설정

Feat: 주문 요청 레디스에 저장 후 merchantUid 발행
- Order: merchantUid, createdAt, productIdList, optionIdList를 가지고 있음. (추가적인 요청이 있을 시 totalMoney도 추가 가능)
- OrderController: 주문을 받을 경우의 api, merchantUid로 주문을 조회하는 api 개발
- OrderService: acceptOrder시 merchantUid를 "ORD-20210815-000005" 형식으로 순차적으로 발행후 레디스에 저장하고 Dto로 반환, getOrder시 merchantUid로 Order를 찾아서 Dto로 반환 

Exception 형식 맞추기
- BaseException을 제공하여 이를 상속하여 Exception 구현할 수 있도록 하였음
- ErrorCode를 enum으로 제공하여 에러코드를 커스텀하여 제공할 수 있도록 하였음
- 향후 Exception처리 시에 이용하기.

ToDo
- OrderService의 ToDo: 현재는 직접 Entity -> Dto매핑을 했는데, Mapstruct같은 것을 적용하여 변경할 것
- ToDo: MySQL 확정 후, Async하게 MySQL에 저장
- ToDo: MySQL에 완전히 저장하는 기능 완성되면, Redis에서 TTL을 주고 MySQL에서 영속화할지 Redis에서 그대로 영속화할지 논의
- Exception의 Todo: ExceptionHandler를 이용하여 적절히 Error 메시지를 api로 묶어서 제공하는 방법 건의

alertjjm committed:
Feat: Redis 설정

- gradle에서 spring data redis 설정
- RedisConfig 설정
- application.properties에 레디스 정보 설정

alertjjm committed:
Feat: 주문 요청 레디스에 저장 후 merchantUid 발행

- ToDo: MySQL 확정 후, Async하게 MySQL에 저장
- ToDo: MySQL에 완전히 저장하는 기능 완성되면, Redis에서 TTL을 주고 MySQL에서 영속화할지 Redis에서 그대로 영속화할지 논의

ssseungzz commented on src/main/kotlin/mashup/sideproject/orderpay/config/RedisConfig.kt:
```diff
@@ -21,24 +20,31 @@ class RedisConfig {
     private val redisPort: Int = 0
 
     @Value("\${spring.redis.password}")
```
이렇게 @value 어노테이션으로 가져오는 것도 괜찮지만 설정할 게 늘어나는 경우에 설정 파일을 데이터 객체로 만드는 것도 괜찮을 것 같아요~! 그런 객체가 만들어져 있을 것 같기도 한데 제가 레디스는 안써봐서 잘 모르겠네요 참고만 해주세용

alertjjm replied:
오 그거도 좋은방법이네오 ㅎㅎㅎㅎ클래스로 따로 빼서 컴포지션해서써도 깔끔하겠네요

KJongHyun commented on src/main/kotlin/mashup/sideproject/orderpay/service/OrderService.kt:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    private fun CharSequence.padStart(length: Int, padChar: Char = ' '): CharSequence {
+        if (length < 0)
```
이 확장함수는 어떤 기능일까요?

KJongHyun replied:
자릿수를 고정하고 0으로 채워주는 기능인가보네요 이러한 기능은 Decimalformat 사용하면 쉽게 처리할 수 있을것 같아요
추가로 이러한 형태의 merchantUid 생성 방법보다는 개인적으로 UUID 사용하는게 좋아 보여요 위 로직에서는 동시에 주문요청이 오면 같은 Uid가 생성될것 같네요

KJongHyun commented on src/main/kotlin/mashup/sideproject/orderpay/config/RedisConfig.kt:
```diff
@@ -0,0 +1,50 @@
+    @Value("\${spring.redis.password}")
+    private val redisPassword: String? = null
+
+    @Bean()
```
@Bean 어노테이션에 특정 옵션이 필요한게아니면 "()"는 생략해도 될것 같아요

alertjjm replied:
앗 실수했네요 감사합니다 ㅎㅎ

KJongHyun commented on src/main/kotlin/mashup/sideproject/orderpay/config/RedisConfig.kt:
```diff
@@ -0,0 +1,50 @@
+        return JedisConnectionFactory(redisStandaloneConfiguration)
+    }
+
+    @Bean("redisTemplateWithJedis")
```
여기도 Bean 어노테이션 안에 value는 생략해도 될것 같아요

alertjjm replied:
넵 그렇네요 ㅎㅎㅎ수정하겠습니다~

KJongHyun commented on src/main/kotlin/mashup/sideproject/orderpay/config/RedisConfig.kt:
```diff
@@ -0,0 +1,50 @@
+    @Bean("redisTemplateWithJedis")
+    fun redisTemplateWithJedis(
+
+    ): RedisTemplate<*, *> {
```
RedisTemplate의 generic key value를 선언과 동일하게 String, Any로 하는게 좋을것 같아요

alertjjm replied:
아 자바에서도 습관적으로 '?' 으로 했었는데 String, Any로 명확하게 하겠습니다 ㅎㅎ

KJongHyun commented on src/main/kotlin/mashup/sideproject/orderpay/controller/OrderController.kt:
```diff
@@ -0,0 +1,24 @@
+@RestController
+class OrderController(private val orderService: OrderService) {
+    @PostMapping
+    fun requestOrder(@RequestBody orderDto: OrderRequestDto?): ResponseEntity<OrderResponseDto?>? {
```
nullable은 필요한곳에만 사용하는게 좋아보입니다!
```
fun requestOrder(@RequestBody orderDto: OrderRequestDto): ResponseEntity<OrderResponseDto>
```

alertjjm replied:
코틀린 nullable이 아직 어렵네요 ..ㅎㅎㅎ 수정하겠습니다 ㅎㅎ nullable더 공부하고 써야겠어요

KJongHyun commented on src/main/kotlin/mashup/sideproject/orderpay/exception/ErrorCode.kt:
```diff
@@ -0,0 +1,20 @@
+) {
+    // Common
+    INVALID_INPUT_VALUE(400, "Invalid Input Value"),
+    METHOD_NOT_ALLOWED(405, "Method Not Allowed"),
```
code는 HttpStatus로 잘 정의된 enum이 있으니 그걸 사용하는게 좋아보여요~~!
ex METHOD_NOT_ALLOWED(HttpStatus.METHOD_NOT_ALLOWED.value(), "Invalid Input Value")

alertjjm replied:
오 숫자로 그냥 적는거보다 종현님이 말씀해주신 방식이 더 좋아보입니다 감사합니다!

KJongHyun commented on src/main/kotlin/mashup/sideproject/orderpay/model/dto/OrderRequestDto.kt:
```diff
@@ -0,0 +1,8 @@
+import java.util.*
+
+data class OrderRequestDto(
+        val productIdList: List<Long>? = ArrayList(),
```
이부분도 이미 ArrayList()로 초기화를 해주고 있기때문에 불필요한 nullable 변수인것 같아요

alertjjm replied:
아.. productIdList가 null이면 ArrayList로 초기화하라는 의미였는데
?=없이 바로 = ArrayList()로 해줘도 될거같네요! 감사합니다

KJongHyun commented on src/main/kotlin/mashup/sideproject/orderpay/config/RedisConfig.kt:
```diff
@@ -0,0 +1,50 @@
+@Configuration
+@EnableRedisRepositories
+class RedisConfig {
+    @Value("\${spring.redis.host}")
```
프로퍼티에서 값을 주입받는 @Value 어노테이션도 생성자에 필드로 동시 선언이 가능해요~

alertjjm replied:
class RedisConfig(@Value(something) host: String, @Value(something2) port: Int) 이런 방식 말씀이신가요?

KJongHyun commented on src/main/kotlin/mashup/sideproject/orderpay/config/RedisConfig.kt:
```diff
@@ -0,0 +1,50 @@
+
+    ): RedisTemplate<*, *> {
+
+        val template = RedisTemplate<String, Any>()
```
kotlin에서 제공하는 scope function 사용하면 조금더 kotlin 스럽게 작성이 가능할것 같아요~

이 부분 이외에도 다른곳에서도 공통으로 참고하시면 좋을것 같네요~
https://kotlinlang.org/docs/scope-functions.html

alertjjm replied:
scope function에 대해서 잘몰라서 찾아보고 수정해보겠습니다!

KJongHyun commented on src/main/kotlin/mashup/sideproject/orderpay/service/OrderService.kt:
```diff
@@ -0,0 +1,63 @@
+
+    private fun createMerchantUid(order: Order): String {
+        val count = orderRepository.count()
+        return "ORD-" +
```
코틀린에서는 문자열 치환을 간단하게 할 수 있습니다
ex
1) "ORD-$count"
2) "ORD-${order.createdAt.format(DateTimeFormatter.ofPattern("yyyyMMdd"))}"

alertjjm replied:
오.. 감사합니다..!!

KJongHyun commented on src/main/kotlin/mashup/sideproject/orderpay/controller/OrderController.kt:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    @GetMapping
+    fun getOrder(@RequestParam merchantUid: String?): ResponseEntity<OrderResponseDto?>? {
```
이 부분은 협의가 필요하긴한데 개인적으로는 2xx status를 세분화해서 사용할게 아니면 Dto를 바로 return하는걸 좋아하긴해요 ㅋㅋ 
그냥 Dto 내리면 200으로 떨어지기때문에

alertjjm replied:
아하 그러면 ResponseEntity로 안묶고 바로 OrderResponseDto로 반환하는것 말씀이시죠?

alertjjm committed:
Fix: 코드리뷰 반영하여 수정

- 에러코드 HttpStatus 사용
- ResponseEntity를 벗김
- 불필요한 nullable 제외
- RedisConfig.kt 리팩토링

alertjjm committed:
Feat: MerchantUid를 UUID 방식으로 변경

alertjjm commented:
![image](https://user-images.githubusercontent.com/46064193/129897853-f37cd954-deac-41e4-a99e-b8744d226c54.png)

UUID방식으로 개선한 모습

alertjjm commented on src/main/kotlin/mashup/sideproject/orderpay/model/entity/Order.kt:
```diff
@@ -0,0 +1,19 @@
+        val id: String? = null,
+        var productIdList: List<Long>? = ArrayList(),
+        var optionIdList: List<Long>? = ArrayList(),
+        @Indexed
```
Member ID, Total Money
